src/vs/editor/contrib/unicodeHighlighter/unicodeHighlighter.ts

```diff
@@ -0,0 +1,504 @@
+			if (!highlightInfo) {
+				continue;
+			}
+			const char = model.getValueInRange(d.range);
+			// text refers to a single character.
+			const codePoint = char.codePointAt(0)!;
+
+			function formatCodePoint(codePoint: number) {
+				return `\`U+${codePoint.toString(16)}\` "${`${escapedMarkdownInlineCode(String.fromCodePoint(codePoint))}`}"`;
+			}
+
+			const codePointStr = formatCodePoint(codePoint);
+
+			let reason: string;
+			switch (highlightInfo.reason.kind) {
+				case 'ambiguous':
+					reason = nls.localize(
+						'unicodeHighlight.characterIsAmbiguous',
+						'The character {0} could be confused with the more common character {1}.',
+						codePointStr,
+						formatCodePoint(highlightInfo.reason.confusableWith.codePointAt(0)!)
+					);
+					break;
+
+				case 'invisible':
+					reason = nls.localize(
+						'unicodeHighlight.characterIsInvisible',
+						'The character {0} is invisible.',
+						codePointStr
+					);
+					break;
+
+				case 'nonBasicAscii':
+					reason = nls.localize(
+						'unicodeHighlight.characterIsNonBasicAscii',
+						'The character {0} is not a basic ASCII character.',
+						codePoint
+					);
+					break;
+			}
+
+			const adjustSettingsArgs: ShowExcludeOptionsArgs = {
+				codePoint: codePoint,
+				reason: highlightInfo.reason.kind,
+			};
+
+			const adjustSettings = nls.localize('unicodeHighlight.adjustSettings', 'Adjust settings');
+			const contents: Array<IMarkdownString> = [{
+				value: `${reason} [${adjustSettings}](command:${ShowExcludeOptions.ID}?${encodeURIComponent(JSON.stringify(adjustSettingsArgs))})`,
+				isTrusted: true,
+			}];
+
+			result.push(new MarkdownHover(this, d.range, contents));
+		}
+		return result;
+	}
+
+	public renderHoverParts(hoverParts: MarkdownHover[], fragment: DocumentFragment, statusBar: IEditorHoverStatusBar): IDisposable {
+		return renderMarkdownHovers(hoverParts, fragment, this._editor, this._hover, this._modeService, this._openerService);
+	}
+}
+
+function escapedMarkdownInlineCode(code: string): string {
+	if (code.indexOf('`') === -1) {
+		return `\`${code}\``;
+	} else {
+		// This is madness. Backticks cannot be escaped in markdown inline code blocks.
+		// Instead, the backtick to start/end an inline code block needs to be repeated so
+		// that the code does not contain a sequence of backticks that is equally long or longer.
+		const delimiter = '`'.repeat(code.length + 1);
+		return `${delimiter} ${code} ${delimiter}`;
+	}
+}
+
+function computeReason(char: string, options: UnicodeHighlighterOptions): UnicodeHighlighterReason | null {
+	return UnicodeTextModelHighlighter.computeUnicodeHighlightReason(char, options);
+}
+
+const DECORATION_HIDE_IN_COMMENTS = ModelDecorationOptions.register({
+	description: 'unicode-highlight',
+	stickiness: TrackedRangeStickiness.NeverGrowsWhenTypingAtEdges,
+	className: 'unicode-highlight',
+	showIfCollapsed: true,
+	overviewRuler: {
+		color: themeColorFromId(overviewRulerFindMatchForeground),
```

Comment:
We could define special colors for this, I just used the find match colors in a hurry.